Parse error when comment appears between variable declaration and #if

### Is it reproducible with SwiftPM command-line tools: `swift build`, `swift test`, `swift package` etc?

- [x] Confirmed reproduction steps with SwiftPM CLI. The description text _must_ include reproduction steps with either of command-line SwiftPM commands, `swift build`, `swift test`, `swift package` etc.

### Description

https://github.com/koher/swift-image.git

MacOS

The following package does not build using the 'swifbuild' backend but does with 'native'

```
swift build `--buildsystem=swiftbuild` fails to build package
```

```
info: Planning Swift module SwiftImageTests (arm64)
    cd /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8
    builtin-SwiftDriver -- /var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/latest.xctoolchain/usr/bin/swiftc -parse-as-library -module-name SwiftImageTests -Onone -enforce-exclusivity\=checked @/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/Objects-normal/arm64/SwiftImageTests.SwiftFileList -DSWIFT_PACKAGE -DDEBUG -DSWIFT_MODULE_RESOURCE_BUNDLE_AVAILABLE -v -DXcode -I /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/swift-DEVELOPMENT-SNAPSHOT-2025-12-01-a-osx/usr/lib/swift/macosx/testing -L /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/swift-DEVELOPMENT-SNAPSHOT-2025-12-01-a-osx/usr/lib/swift/macosx/testing -plugin-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/swift-DEVELOPMENT-SNAPSHOT-2025-12-01-a-osx/usr/lib/swift/host/plugins/testing -sdk /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX16.0.sdk -F /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -F /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/PrivateFrameworks -I /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib -L /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib -g -external-plugin-path /var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/latest.xctoolchain/usr/lib/swift/host/plugins/testing\#/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/latest.xctoolchain/usr/bin/swift-plugin-server -enable-experimental-feature DebugDescriptionMacro -sdk /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX16.0.sdk -target arm64-apple-macos14.0 -g -module-cache-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/ModuleCache.noindex -Xfrontend -serialize-debugging-options -enable-testing -index-store-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx -Xcc -D_LIBCPP_HARDENING_MODE\=_LIBCPP_HARDENING_MODE_DEBUG -swift-version 5 -I /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Products/Debug -Isystem /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib -Isystem /var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/latest.xctoolchain/usr/lib/swift/macosx/testing -F /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Products/Debug/PackageFrameworks -F /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Products/Debug -F /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -c -j24 -enable-batch-mode -incremental -output-file-map /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/Objects-normal/arm64/SwiftImageTests-OutputFileMap.json -use-frontend-parseable-output -save-temps -no-color-diagnostics -explicit-module-build -module-cache-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftExplicitPrecompiledModules -clang-scanner-module-cache-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/ModuleCache.noindex -sdk-module-cache-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/ModuleCache.noindex -serialize-diagnostics -emit-dependencies -emit-module -emit-module-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/Objects-normal/arm64/SwiftImageTests.swiftmodule -validate-clang-modules-once -clang-build-session-file /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/ModuleCache.noindex/Session.modulevalidation -package-name swift_image -emit-const-values -Xfrontend -const-gather-protocols-file -Xfrontend /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/Objects-normal/arm64/SwiftImageTests_const_extract_protocols.json -Xcc -I/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Products/Debug/include -Xcc -I/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/DerivedSources-normal/arm64 -Xcc -I/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/DerivedSources/arm64 -Xcc -I/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/DerivedSources -Xcc -DSWIFT_PACKAGE -Xcc -DDEBUG\=1 -working-directory /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8 -experimental-emit-module-separately -disable-cmo
/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/Tests/SwiftImageTests/SwiftImageTests.swift:17:23: error: consecutive statements on a line must be separated by ';'
        /**/ let y = 0
                      ^
                      ;
/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/Tests/SwiftImageTests/SwiftImageTests.swift:40:66: error: consecutive statements on a line must be separated by ';'
        let binarized: Image<Bool> = image.map { $0.gray >= 127 }
                                                                 ^
                                                                 ;
Apple Swift version 6.3-dev (LLVM 35f48306184cd25, Swift a69dbb3366e64d8)
Target: arm64-apple-macos14.0
error: /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/Tests/SwiftImageTests/SwiftImageTests.swift:17:23 consecutive statements on a line must be separated by ';'
        /**/ let y = 0
                      ^
                      ;
error: /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/Tests/SwiftImageTests/SwiftImageTests.swift:40:66 consecutive statements on a line must be separated by ';'
        let binarized: Image<Bool> = image.map { $0.gray >= 127 }
                                                                 ^
                                                                 ;
error: SwiftDriver SwiftImageTests normal arm64 com.apple.xcode.tools.swift.compiler failed with a nonzero exit code. Command line:     cd /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8
    builtin-SwiftDriver -- /var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/latest.xctoolchain/usr/bin/swiftc -parse-as-library -module-name SwiftImageTests -Onone -enforce-exclusivity\=checked @/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/Objects-normal/arm64/SwiftImageTests.SwiftFileList -DSWIFT_PACKAGE -DDEBUG -DSWIFT_MODULE_RESOURCE_BUNDLE_AVAILABLE -v -DXcode -I /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/swift-DEVELOPMENT-SNAPSHOT-2025-12-01-a-osx/usr/lib/swift/macosx/testing -L /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/swift-DEVELOPMENT-SNAPSHOT-2025-12-01-a-osx/usr/lib/swift/macosx/testing -plugin-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/swift-DEVELOPMENT-SNAPSHOT-2025-12-01-a-osx/usr/lib/swift/host/plugins/testing -sdk /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX16.0.sdk -F /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -F /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/PrivateFrameworks -I /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib -L /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib -g -external-plugin-path /var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/latest.xctoolchain/usr/lib/swift/host/plugins/testing\#/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/latest.xctoolchain/usr/bin/swift-plugin-server -enable-experimental-feature DebugDescriptionMacro -sdk /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX16.0.sdk -target arm64-apple-macos14.0 -g -module-cache-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/ModuleCache.noindex -Xfrontend -serialize-debugging-options -enable-testing -index-store-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx -Xcc -D_LIBCPP_HARDENING_MODE\=_LIBCPP_HARDENING_MODE_DEBUG -swift-version 5 -I /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Products/Debug -Isystem /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib -Isystem /var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/swift-toolchains/latest.xctoolchain/usr/lib/swift/macosx/testing -F /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Products/Debug/PackageFrameworks -F /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Products/Debug -F /Applications/XcodeWonderXIP.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -c -j24 -enable-batch-mode -incremental -output-file-map /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/Objects-normal/arm64/SwiftImageTests-OutputFileMap.json -use-frontend-parseable-output -save-temps -no-color-diagnostics -explicit-module-build -module-cache-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftExplicitPrecompiledModules -clang-scanner-module-cache-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/ModuleCache.noindex -sdk-module-cache-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/ModuleCache.noindex -serialize-diagnostics -emit-dependencies -emit-module -emit-module-path /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/Objects-normal/arm64/SwiftImageTests.swiftmodule -validate-clang-modules-once -clang-build-session-file /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/ModuleCache.noindex/Session.modulevalidation -package-name swift_image -emit-const-values -Xfrontend -const-gather-protocols-file -Xfrontend /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/Objects-normal/arm64/SwiftImageTests_const_extract_protocols.json -Xcc -I/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Products/Debug/include -Xcc -I/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/DerivedSources-normal/arm64 -Xcc -I/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/DerivedSources/arm64 -Xcc -I/private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8/swift-image/.build-swiftbuild/arm64-apple-macosx/Intermediates.noindex/SwiftImage.build/Debug/SwiftImageTests.build/DerivedSources -Xcc -DSWIFT_PACKAGE -Xcc -DDEBUG\=1 -working-directory /private/var/lib/jenkins/workspace/swift_oss/swift-package-index-metrics/spi-workdir/000DBA3D-CE2C-4681-A888-7417DF0087E8 -experimental-emit-module-separately -disable-cmo
error: Build failed

```

### Expected behaviour

Builds

### Actual behavior

Fails to build

### Steps to reproduce

```
git clone https://github.com/koher/swift-image.git
swift build --build-system=swiftbuild
```

### Swift Package Manager version/commit hash

_No response_

### Swift & OS version (output of `swift --version ; uname -a`)

_No response_

--- a/include/swift/Parse/Lexer.h
+++ b/include/swift/Parse/Lexer.h
@@ -39,7 +39,6 @@ namespace swift {
   template<typename ...T> struct Diag;
 
 enum class CommentRetentionMode {
-  None,
   AttachToNextToken,
   ReturnAsTokens,
 };
@@ -185,7 +184,7 @@ public:
       const LangOptions &Options, const SourceManager &SourceMgr,
       unsigned BufferID, DiagnosticEngine *Diags, LexerMode LexMode,
       HashbangMode HashbangAllowed = HashbangMode::Disallowed,
-      CommentRetentionMode RetainComments = CommentRetentionMode::None);
+      CommentRetentionMode RetainComments = CommentRetentionMode::AttachToNextToken);
 
   /// Create a lexer that scans a subrange of the source buffer.
   Lexer(const LangOptions &Options, const SourceManager &SourceMgr,

--- a/include/swift/Parse/Parser.h
+++ b/include/swift/Parse/Parser.h
@@ -693,6 +693,10 @@ public:
     return Context.LangOpts.EnableExperimentalConcurrency;
   }
 
+  bool shouldAttachCommentToDecl() {
+    return Context.LangOpts.AttachCommentsToDecls;
+  }
+
   /// Returns true if a Swift declaration starts after the current token,
   /// otherwise returns false.
   bool isNextStartOfSwiftDecl() {

--- a/lib/IDE/Formatting.cpp
+++ b/lib/IDE/Formatting.cpp
@@ -49,7 +49,7 @@ static void widenOrSet(SourceRange &First, SourceRange Second) {
 static bool isFirstTokenOnLine(SourceManager &SM, SourceLoc Loc) {
   assert(Loc.isValid());
   SourceLoc LineStart = Lexer::getLocForStartOfLine(SM, Loc);
-  CommentRetentionMode SkipComments = CommentRetentionMode::None;
+  CommentRetentionMode SkipComments = CommentRetentionMode::AttachToNextToken;
   Token First = Lexer::getTokenAtLocation(SM, LineStart, SkipComments);
   return First.getLoc() == Loc;
 }
@@ -76,8 +76,8 @@ static std::optional<Token> getTokenAfter(SourceManager &SM, SourceLoc Loc,
                                           bool SkipComments = true) {
   assert(Loc.isValid());
   CommentRetentionMode Mode = SkipComments
-    ? CommentRetentionMode::None
-    : CommentRetentionMode::ReturnAsTokens;
+                                  ? CommentRetentionMode::AttachToNextToken
+                                  : CommentRetentionMode::ReturnAsTokens;
   assert(Lexer::getTokenAtLocation(SM, Loc, Mode).getLoc() == Loc);
   SourceLoc End = Lexer::getLocForEndOfToken(SM, Loc);
   Token Next = Lexer::getTokenAtLocation(SM, End, Mode);
@@ -814,8 +814,8 @@ class OutdentChecker: protected RangeWalker {
     } else if (R.isValid()) {
       // Check condition 2a.
       SourceLoc LineStart = Lexer::getLocForStartOfLine(SM, R);
-      Token First = Lexer::getTokenAtLocation(SM, LineStart,
-                                              CommentRetentionMode::None);
+      Token First = Lexer::getTokenAtLocation(
+          SM, LineStart, CommentRetentionMode::AttachToNextToken);
       IsOutdenting |= First.getLoc() == R;
     }
 

--- a/lib/Parse/Lexer.cpp
+++ b/lib/Parse/Lexer.cpp
@@ -270,7 +270,7 @@ Token Lexer::getTokenAt(SourceLoc Loc) {
          "location from the wrong buffer");
 
   Lexer L(LangOpts, SourceMgr, BufferID, getUnderlyingDiags(), LexMode,
-          HashbangMode::Allowed, CommentRetentionMode::None);
+          HashbangMode::Allowed, CommentRetentionMode::AttachToNextToken);
   L.restoreState(State(Loc));
   return L.peekNextToken();
 }
@@ -3029,7 +3029,7 @@ static SourceLoc getLocForStartOfTokenInBuf(SourceManager &SM,
   LangOptions FakeLangOptions;
 
   Lexer L(FakeLangOptions, SM, BufferID, nullptr, LexerMode::Swift,
-          HashbangMode::Allowed, CommentRetentionMode::None,
+          HashbangMode::Allowed, CommentRetentionMode::AttachToNextToken,
           BufferStart, BufferEnd);
 
   // Lex tokens until we find the token that contains the source location.

--- a/lib/Parse/ParseDecl.cpp
+++ b/lib/Parse/ParseDecl.cpp
@@ -6204,7 +6204,7 @@ ParserStatus Parser::parseDecl(bool IsAtStartOfLineOrPreviousHadSemi,
 
   // Parse attributes.
   DeclAttributes Attributes;
-  if (Tok.hasComment())
+  if (Tok.hasComment() && shouldAttachCommentToDecl())
     Attributes.add(new (Context) RawDocCommentAttr(Tok.getCommentRange()));
   ParserStatus AttrStatus = parseDeclAttributeList(
       Attributes, IfConfigsAreDeclAttrs);

--- a/lib/Parse/ParseGeneric.cpp
+++ b/lib/Parse/ParseGeneric.cpp
@@ -56,7 +56,7 @@ Parser::parseGenericParametersBeforeWhere(SourceLoc LAngleLoc,
 
     // Parse attributes.
     DeclAttributes attributes;
-    if (Tok.hasComment())
+    if (Tok.hasComment() && shouldAttachCommentToDecl())
       attributes.add(new (Context) RawDocCommentAttr(Tok.getCommentRange()));
     parseDeclAttributeList(attributes);
 

--- a/lib/Parse/Parser.cpp
+++ b/lib/Parse/Parser.cpp
@@ -341,21 +341,15 @@ Parser::Parser(unsigned BufferID, SourceFile &SF, SILParserStateBase *SIL,
                PersistentParserState *PersistentState)
     : Parser(BufferID, SF, &SF.getASTContext().Diags, SIL, PersistentState) {}
 
-Parser::Parser(unsigned BufferID, SourceFile &SF, DiagnosticEngine* LexerDiags,
-               SILParserStateBase *SIL,
-               PersistentParserState *PersistentState)
-    : Parser(
-          std::unique_ptr<Lexer>(new Lexer(
-              SF.getASTContext().LangOpts, SF.getASTContext().SourceMgr,
-              BufferID, LexerDiags,
-              sourceFileKindToLexerMode(SF.Kind),
-              SF.Kind == SourceFileKind::Main
-                  ? HashbangMode::Allowed
-                  : HashbangMode::Disallowed,
-              SF.getASTContext().LangOpts.AttachCommentsToDecls
-                  ? CommentRetentionMode::AttachToNextToken
-                  : CommentRetentionMode::None)),
-          SF, SIL, PersistentState) {}
+Parser::Parser(unsigned BufferID, SourceFile &SF, DiagnosticEngine *LexerDiags,
+               SILParserStateBase *SIL, PersistentParserState *PersistentState)
+    : Parser(std::unique_ptr<Lexer>(new Lexer(
+                 SF.getASTContext().LangOpts, SF.getASTContext().SourceMgr,
+                 BufferID, LexerDiags, sourceFileKindToLexerMode(SF.Kind),
+                 SF.Kind == SourceFileKind::Main ? HashbangMode::Allowed
+                                                 : HashbangMode::Disallowed,
+                 CommentRetentionMode::AttachToNextToken)),
+             SF, SIL, PersistentState) {}
 
 namespace {
 
@@ -1243,12 +1237,10 @@ ParserUnit::ParserUnit(SourceManager &SM, SourceFileKind SFKind,
     : Impl(*new Implementation(SM, SFKind, BufferID, LangOptions(), "input")) {
 
   std::unique_ptr<Lexer> Lex;
-  Lex.reset(new Lexer(Impl.LangOpts, SM,
-                      BufferID, &Impl.Diags,
-                      LexerMode::Swift,
-                      HashbangMode::Allowed,
-                      CommentRetentionMode::None,
-                      Offset, EndOffset));
+  Lex.reset(new Lexer(Impl.LangOpts, SM, BufferID, &Impl.Diags,
+                      LexerMode::Swift, HashbangMode::Allowed,
+                      CommentRetentionMode::AttachToNextToken, Offset,
+                      EndOffset));
   Impl.TheParser.reset(new Parser(std::move(Lex), *Impl.SF, /*SIL=*/nullptr,
                                   /*PersistentState=*/nullptr));
 }

--- a/lib/Refactoring/Async/AsyncConverter.cpp
+++ b/lib/Refactoring/Async/AsyncConverter.cpp
@@ -651,8 +651,8 @@ void AsyncConverter::addFuncDecl(const FuncDecl *FD) {
     LeftEndLoc = Lexer::getLocForEndOfToken(
         SM, Params->get(TopHandler.Index - 1)->getEndLoc());
     // Skip to the end of any comments
-    Token Next =
-        Lexer::getTokenAtLocation(SM, LeftEndLoc, CommentRetentionMode::None);
+    Token Next = Lexer::getTokenAtLocation(
+        SM, LeftEndLoc, CommentRetentionMode::AttachToNextToken);
     if (Next.getKind() != tok::NUM_TOKENS)
       LeftEndLoc = Next.getLoc();
     break;

--- a/lib/Sema/MiscDiagnostics.cpp
+++ b/lib/Sema/MiscDiagnostics.cpp
@@ -2559,8 +2559,9 @@ public:
     // insert 'self,'. If it wasn't a valid entry, then we will at least not
     // be introducing any new errors/warnings...
     const auto locAfterBracket = brackets.Start.getAdvancedLoc(1);
-    const auto nextAfterBracket = Lexer::getTokenAtLocation(
-        Ctx.SourceMgr, locAfterBracket, CommentRetentionMode::None);
+    const auto nextAfterBracket =
+        Lexer::getTokenAtLocation(Ctx.SourceMgr, locAfterBracket,
+                                  CommentRetentionMode::AttachToNextToken);
     if (nextAfterBracket.getLoc() != brackets.End)
       diag.fixItInsertAfter(brackets.Start, "self, ");
     else
@@ -2581,8 +2582,8 @@ public:
     // opening brace of the closure, we may need to pad the fix-it
     // with a space.
     const auto nextLoc = closureExpr->getLoc().getAdvancedLoc(1);
-    const auto next = Lexer::getTokenAtLocation(Ctx.SourceMgr, nextLoc,
-                                                CommentRetentionMode::None);
+    const auto next = Lexer::getTokenAtLocation(
+        Ctx.SourceMgr, nextLoc, CommentRetentionMode::AttachToNextToken);
     std::string trailing = next.getLoc() == nextLoc ? " " : "";
 
     diag.fixItInsertAfter(closureExpr->getLoc(), " [self] in" + trailing);

--- a/test/Parse/backgtrack-startofline.swift
+++ b/test/Parse/backgtrack-startofline.swift
@@ -1,0 +1,7 @@
+// RUN: %target-typecheck-verify-swift -verify-ignore-unrelated
+
+func test() {
+  _ = "test"
+  /* comment */ #if true
+  #endif
+}

--- a/unittests/Parse/LexerTests.cpp
+++ b/unittests/Parse/LexerTests.cpp
@@ -271,32 +271,6 @@ TEST_F(LexerTest, ContentStartTokenIsStartOfLineUTF8BOM) {
   ASSERT_TRUE(Tok.isAtStartOfLine());
 }
 
-TEST_F(LexerTest, BOMNoCommentNoTrivia) {
-  const char *Source = "\xEF\xBB\xBF" "// comment\naaa //xx \n/* x */";
-  
-  LangOptions LangOpts;
-  SourceManager SourceMgr;
-  unsigned BufferID = SourceMgr.addMemBufferCopy(StringRef(Source));
-  
-  Lexer L(LangOpts, SourceMgr, BufferID, /*Diags=*/nullptr, LexerMode::Swift,
-          HashbangMode::Disallowed, CommentRetentionMode::None);
-  
-  Token Tok;
-
-  L.lex(Tok);
-  ASSERT_EQ(tok::identifier, Tok.getKind());
-  ASSERT_EQ("aaa", Tok.getText());
-  ASSERT_EQ(SourceMgr.getLocForOffset(BufferID, 14), Tok.getLoc());
-  ASSERT_EQ(SourceMgr.getLocForOffset(BufferID, 14), Tok.getCommentRange().getStart());
-  ASSERT_EQ(0u, Tok.getCommentRange().getByteLength());
-  
-  L.lex(Tok);
-  ASSERT_EQ(tok::eof, Tok.getKind());
-  ASSERT_EQ(SourceMgr.getLocForOffset(BufferID, 31), Tok.getLoc());
-  ASSERT_EQ(SourceMgr.getLocForOffset(BufferID, 31), Tok.getCommentRange().getStart());
-  ASSERT_EQ(0u, Tok.getCommentRange().getByteLength());
-}
-
 TEST_F(LexerTest, BOMTokenCommentNoTrivia) {
   const char *Source = "\xEF\xBB\xBF" "// comment\naaa //xx \n/* x */";
   
@@ -718,9 +692,8 @@ TEST_F(LexerTest, DiagnoseEmbeddedNul) {
   DiagnosticEngine Diags(SourceMgr);
   Diags.addConsumer(DiagConsumer);
 
-  Lexer L(LangOpts, SourceMgr, BufferID, &Diags,
-          LexerMode::Swift, HashbangMode::Disallowed,
-          CommentRetentionMode::None);
+  Lexer L(LangOpts, SourceMgr, BufferID, &Diags, LexerMode::Swift,
+          HashbangMode::Disallowed, CommentRetentionMode::AttachToNextToken);
 
   Token Tok;
   L.lex(Tok);
@@ -743,9 +716,8 @@ TEST_F(LexerTest, DiagnoseEmbeddedNulOffset) {
   DiagnosticEngine Diags(SourceMgr);
   Diags.addConsumer(DiagConsumer);
 
-  Lexer L(LangOpts, SourceMgr, BufferID, &Diags,
-          LexerMode::Swift, HashbangMode::Disallowed,
-          CommentRetentionMode::None,
+  Lexer L(LangOpts, SourceMgr, BufferID, &Diags, LexerMode::Swift,
+          HashbangMode::Disallowed, CommentRetentionMode::AttachToNextToken,
           /*Offset=*/5, /*EndOffset=*/SourceLen);
 
   ASSERT_FALSE(containsPrefix(

